go-controller/pkg/cni/cniserver.go

```diff
@@ -51,15 +56,45 @@ func NewCNIServer(rundir string, kclient kubernetes.Interface) *Server {
 		Server: http.Server{
 			Handler: router,
 		},
-		rundir:  rundir,
-		kclient: kclient,
+		rundir:    rundir,
+		podLister: corev1listers.NewPodLister(factory.LocalPodInformer().GetIndexer()),
+		adds:      make(map[string]*PodRequest),
 	}
 	router.NotFoundHandler = http.HandlerFunc(http.NotFound)
 	router.HandleFunc("/", s.handleCNIRequest).Methods("POST")
 	router.HandleFunc("/metrics", s.handleCNIMetrics).Methods("POST")
+
+	factory.AddPodHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*kapi.Pod)
+			s.cancelEarliestPodAdd(pod)
+		},
+	}, nil)
+
 	return s
 }
 
+// cancelEarliestPodAdd requests that the earliest outstanding add operation for a given
```

Comment:
`Earliest` or `earlier than`? 
There should be only on ADD active, no?

```
	for _, req := range s.adds {
		if req.PodNamespace == pod.Namespace && req.PodName == pod.Name && req.timestamp.Before(earliest) {
			req.cancel()
		}
	}

Reply:
There may be >= 0 sandboxes for the same Namespace+Name. When a sandbox fails, or the pod is deleted, the sandbox is put on the GC list and deleted shortly after (up to 3 minutes IIRC). During that time, another sandbox may be started for the same pod, or a new instance of the pod.

Reply:
Since we don't have the Pod UID here, we just assume that the delete should cancel the earliest sandbox, because that sandbox is superceded by the newer one anyway.

Reply:
can you add a comment with that explanation it will be very useful

Reply:
Done.